### PR TITLE
Enable tnt to mutate projectile velocity

### LINKED ISSUE #5429: tnt not affecting projectiles
### Expected behavior

I expected tnt to launch projectiles.

### Observed/Actual behavior

tnt did not launch projectiles, either in or out of water.

### Steps/models to reproduce

place tnt, spawn arrows / any projectile, light tnt

### Plugin list

None

### Paper version

This server is running Paper version git-Paper-571 (MC: 1.16.5) (Implementing API version 1.16.5-R0.1-SNAPSHOT)
You are running the latest version

### Agreements

- [X] I am running the latest version of Paper available from https://papermc.io/downloads.
- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] My version of Minecraft is supported by Paper.

### PR DESCRIPTION
**Pull request by lynxplay**
Spigot implemented a non-vanilla check prior to modifying an entities
velocity due to an explosion that required the Entity#damageEntity call
to return `true` (presumably) to not move entities for which the damage
was cancelled.

While the spigot implementation already exempted primed tnt and falling
blocks from this condition, as they never take damage and always return
`false`, projectiles did not recieve the same treatment. As most
projectiles cannot take damage (as they either simply ignore damage or
die immediately) projectiles such as the arrow would no longer be
affected by an explosion and would hence not change their velocity.

This commit returns the vanilla behaviour by allowing all entities that
extend the `IProjectile` class to change their velocity due to an
explosion by adding an `instanceof` check to the two previously existing
`instanceof` checks for primed tnt and falling block entities.

Resolves: #5429

----

NOTE:
As of right now, this PR now enables **<u>all</u>** entities that extend IProjectile to be moved by tnt. This evaluates to the following entities:
- DragonFireball
- Egg
- EnderPearl
- Fireworks
- FishingHook
- LargeFireball
- Potion
- ShulkerBullet
- SmallFireball
- SpectralArrow
- ThrownExpBottle
- ThrownTrident
- TippedArrow
- WitherSkull

### CODE REVIEW

File: Spigot-Server-Patches/0699-Enable-tnt-to-mutate-projectile-velocity.patch [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+index 79008bda42558ea7d28ccf51b66405a3bdb52da7..2cab2f14a0c089cae1562064e64b1e886b1119bd 100644
+--- a/src/main/java/net/minecraft/world/level/Explosion.java
++++ b/src/main/java/net/minecraft/world/level/Explosion.java
+@@ -223,9 +223,11 @@ public class Explosion {
+                         entity.forceExplosionKnockback = false;
+                         boolean wasDamaged = entity.damageEntity(this.b(), (float) ((int) ((d13 * d13 + d13) / 2.0D * 7.0D * (double) f2 + 1.0D)));
+                         CraftEventFactory.entityDamage = null;
+-                        if (!wasDamaged && !(entity instanceof EntityTNTPrimed || entity instanceof EntityFallingBlock) && !entity.forceExplosionKnockback) {
++                        // Paper start - Allow projectiles to still be moved by explosion without taking damage
++                        if (!wasDamaged && !(entity instanceof EntityTNTPrimed || entity instanceof EntityFallingBlock || entity instanceof IProjectile) && !entity.forceExplosionKnockback) {
```
Comment (Machine-Maker) [inline Spigot-Server-Patches/0699-Enable-tnt-to-mutate-projectile-velocity.patch:35]: This is only changing one line right? Can the comment be moved inline (w/the description still ofc).

Comment (lynxplay) [reply Spigot-Server-Patches/0699-Enable-tnt-to-mutate-projectile-velocity.patch:35]: It definitely can but (as the git hunk does include those lines anyway) I moved it there due to the horrible length of the single line. If paper does not care about the line length then I can obviously change it 👍

Comment (traksag) [reply Spigot-Server-Patches/0699-Enable-tnt-to-mutate-projectile-velocity.patch:35]: One edge case is that `damageEntity` generates damage events for large fireballs (implementation in `EntityFireball`). If a plugin cancels such an explosion damage event, knockback will still be applied to the large fireball with this change in place. Possibly exclude large fireballs from the `instanceof` test to avoid this.

Comment (lynxplay) [reply Spigot-Server-Patches/0699-Enable-tnt-to-mutate-projectile-velocity.patch:35]: The edge case goes even deeper than that tho:
The `EntityFireball` class only ever calls the [EntityDamageByEntityEvent](https://papermc.io/javadocs/paper/1.16/org/bukkit/event/entity/EntityDamageByEntityEvent.html) if the `source` of the explosion that damaged the `EntityFireball` exists.

This means that the event is only called if the explosion can be tracked back to a player (player ignites tnt, projectile launched by player caused an explosion through plugin, etc..). A tnt dispensed by a dispenser does not trigger the [EntityDamageByEntityEvent](https://papermc.io/javadocs/paper/1.16/org/bukkit/event/entity/EntityDamageByEntityEvent.html).

This means that excluding the `EntityFireball` from this patch would result in player-caused explosion being the only explosions that can move the fireball.

Comment (lynxplay) [reply Spigot-Server-Patches/0699-Enable-tnt-to-mutate-projectile-velocity.patch:35]: Looking further into this I don't think we are capable of preventing this without completely changing how the `wasDamanged` variable is set.

This means that this PR either:
1. allows explosions caused directly from the player to mutate a fireballs velocity even if the damage event was cancelled
2. cannot mutate fireball velocity at all (closer to vanilla but does break the EntityDamageByEntity contract about preventing knockback in the specific "fireball caught in player explosion" edgecase).

**Comment by lynxplay** [comment]
I personally don't know if we want to attempt a fix of #5415 in this PR as well, concerning they touch the same piece of code.

The PR will keep, unless specifically requested otherwise, the edge case of `large_fireball` entities being affected by an explosion in regards to their velocity even if the event was cancelled. With the current method signature of `Entity#damageEntity` we are not able to determine whether or not the entity was not damaged because it cannot be damaged at all (see `EntityFireball`) or because the created `EntityDamageEvent` was cancelled.

For the future of this PR I'd personally appreciate feedback on the edge case and feedback on an inclusion of a potential fix for #5415.

**Comment by lynxplay** [comment]
Closed in favour of #5516.
